exporter/metric/prometheus/prometheus.go

```diff
@@ -104,70 +92,163 @@ func NewExporter(opts Options) (*Exporter, error) {
 		opts.Gatherer = opts.Registry
 	}
 
-	return &Exporter{
-		registerer:         opts.Registerer,
-		gatherer:           opts.Gatherer,
-		handler:            promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}),
-		measureAggregation: opts.MeasureAggregation,
+	if opts.DefaultSummaryObjectives == nil {
+		opts.DefaultSummaryObjectives = map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001}
+	}
+
+	if opts.OnError == nil {
+		opts.OnError = func(err error) {
+			fmt.Println(err.Error())
+		}
+	}
+
+	e := &Exporter{
+		handler:                  promhttp.HandlerFor(opts.Gatherer, promhttp.HandlerOpts{}),
+		registerer:               opts.Registerer,
+		gatherer:                 opts.Gatherer,
+		DefaultHistogramBuckets:  opts.DefaultHistogramBuckets,
+		DefaultSummaryObjectives: opts.DefaultSummaryObjectives,
+	}
+
+	c := newCollector(e)
+	if err := opts.Registerer.Register(c); err != nil {
+		opts.OnError(fmt.Errorf("cannot register the collector: %w", err))
+	}
 
-		counters:   newCounters(opts.Registerer),
-		gauges:     newGauges(opts.Registerer),
-		histograms: newHistograms(opts.Registerer, opts.DefaultHistogramBuckets),
-		summaries:  newSummaries(opts.Registerer, opts.DefaultSummaryObjectives),
-	}, nil
+	return e, nil
 }
 
 // Export exports the provide metric record to prometheus.
 func (e *Exporter) Export(_ context.Context, checkpointSet export.CheckpointSet) error {
-	var forEachError error
-	checkpointSet.ForEach(func(record export.Record) {
-		agg := record.Aggregator()
+	e.snapshot = checkpointSet
```

Comment:
It may be safe to do this, but I think it will be required to block collection while this is being read.

Reply:
Yes, I had some concerns about this. `ForEach` method ranges on a map and in the collection phase it could be adding new keys to this map. Do you think synchronizing this with a mutex would be okay for now? We could change this if we move forward to a pull controller and synchronize only there.

Other solutions I've come up was:
- Copying keys on to a slice and iterating over this, but again we might have a problem with reading an export and the collector thread calling the `Merge` method which is stated on the docs that this is called in a "single-thread context";
- Creating the prometheus metrics inside the `Export()` method and keep it on memory. This may result in a higher memory usage but at least should be a little bit better for GC since the current implementation creates a prometheus metric descriptor twice (in `Describe()` and `Collect()`)

We should probably add some docs on whether a `Checkpoint` could be copied and/or consumed later.

Reply:
A Mutex is fine for now. It means that collection will be delayed when a current request is querying, that's fine.